### PR TITLE
Clean up event listener logic, add support for consecutive events with `.then`

### PR DESCRIPTION
**Pull request by aliabid94**
Cleaned up the repeated logic across all event listeners. Created a Callable `EventListenerMethod` and appended it as an attribute in the constructor of the EventListeners.

Fixes: #2560

Also added the ability to chain event listeners, using a `.then()` method. For example:
```python
num1 = gr.Number()
num2 = gr.Number()
num3= gr.Number()
num4 = gr.Number()
btn = gr.Button()

btn.click(add, [num1, num2], num3).then(square, num3, num4)
```

See a more useful example in demo/chatbot_simple_demo.

### CODE REVIEW

**Comment by abidlabs** [comment]
Niceeee

File: gradio/events.py [hunk 1]
```diff
@@ -78,703 +96,123 @@ def change(
...
-        queue=None,
-        batch: bool = False,
-        max_batch_size: int = 4,
-        preprocess: bool = True,
-        postprocess: bool = True,
-        cancels: Dict[str, Any] | List[Dict[str, Any]] | None = None,
-        every: float | None = None,
-        _js: str | None = None,
-    ):
+    def __init__(self):
```
Comment (freddyaboulton) [inline gradio/events.py:132]: The typehints are preserved which is awesome but we're losing the docstring.

### Main
image


### This branch
image


I think a good workaround would be manually adding the docstring of `EventListenerMethod` to the docstring you already have here. For example,

```python

    def __init__(self):
        self.click = EventListenerMethod(self, "click")
        f"""
        This event is triggered when the component (e.g. a button) is clicked.
        This method can be used when this component is in a Gradio Blocks.
        
        {self.click.__doc__}
        """
```

File: gradio/events.py [hunk 2]
```diff
@@ -32,19 +32,40 @@ def set_cancel_events(
...
-    pass
+    def __init__(self):
+        for event_listener_class in EventListener.__subclasses__():
+            if isinstance(self, event_listener_class):
+                event_listener_class.__init__(self)
 
 
-class Changeable(EventListener):
-    def change(
+class Dependency(dict):
```
Comment (aliabid94) [inline gradio/events.py:41]: inherits from dict so event listener return type is backwards compatible

**Review by freddyaboulton — COMMENTED** [review]
I love the clean up @aliabid94 ! Only one comment on that, rest looks good to me.

I'm not 100% sure about `then` though. It works great but I think last time we talked about this, there were some concerns about implementing it. I'll try to dig that up.

Also not sure about having the canonical chatbot "hello world" example use `then`.

File: gradio/events.py (hunk 1 above)
Comment (aliabid94) [reply gradio/events.py:132]: clever! will do

Comment (aliabid94) [reply gradio/events.py:132]: hmm I still didn't get the type hint - does it work for you @freddyaboulton?

**Comment by aliabid94** [comment]
happy to rekindle that conversation! (some of it is in the linked issue).

The chatbot was actually what pushed me to implement `.then`, I didn't like that there was a wait and then both the user and bot respond together.

File: gradio/events.py (hunk 1 above)
Comment (aliabid94) [reply gradio/events.py:132]: btw the behaviour in this branch rn for me is: I get the docstring while typing `click`, and argument type hints when I open the parentheses.

Comment (freddyaboulton) [reply gradio/events.py:132]: Dang yea thought this was working for me. Let me think about this some more, I think the problem is that the `click` is determined at runtime. Should be fine to proceed without this though?

Comment (abidlabs) [reply gradio/events.py:132]: I get the same thing @aliabid94. Docstring while typing in `.click()`: 

image

Type hints after opening the parentheses:

image

Very cool!

**Comment by pngwn** [comment]
Can you add more context to the PR description please. There was a lot of discussion about .then with a number of concerns, I'm not sure if these have been addressed. 

I'm not personally convinced that this is the right approach, I don't think it is a good API. 

Feels like we're just ramming in a bunch of new API atm because it enables something useful for one usecase without actually considering the API that much.

**Comment by pngwn** [comment]
I still don't think this feature is an idealw solution for the problem stated because the demo added doesn't really give normal chatbot functionality. You actually need to set the textbox as an output of the first function too:

```py
def add_message(chat_history, message):
    return [chat_history + [[message, None]], ""]

def respond(chat_history):
    time.sleep(1)
    response = random.choice(["Yes", "No"])
    chat_history[-1][1] = response
    return chat_history

with gr.Blocks() as demo:
    chatbot = gr.Chatbot()
    msg = gr.Textbox()
    clear = gr.Button("Clear")

    msg.submit(add_message, [chatbot, msg], [chatbot, msg]).then(
        respond, chatbot, chatbot
    )
    clear.click(lambda: None, None, chatbot)
```

This still feels like a lot of work to meet the needs of the standard usecase for the chatbot.

**Comment by abidlabs** [comment]
Let's split up the PR, will make it easier to discuss the implications of adding `.then()` to the API. Just to clarify, though, there are definitely lots of concrete use cases for `.then()`. I mention 2 in the original issue, having a proper chatbot is good reason, and there are others that have come up recently I can describe.

**Comment by aliabid94** [comment]
ok will split this up, but let's continue the `.then` discussion in the issue

**Pull request closed without merging**